src/Symfony/Component/Lock/Store/MongoDbStore.php

```diff
@@ -0,0 +1,294 @@
+     * garbage-collection. Alternatively it's possible to automatically expire
+     * the locks in the database as described below:
+     *
+     * A TTL collections can be used on MongoDB 2.2+ to cleanup expired locks
+     * automatically. Such an index can for example look like this:
+     *
+     *     db.<session-collection>.ensureIndex(
+     *         { "<expiry-field>": 1 },
+     *         { "expireAfterSeconds": 0 }
+     *     )
+     *
+     * More details on: http://docs.mongodb.org/manual/tutorial/expire-data/
+     * @param float $initialTtl The expiration delay of locks in seconds
+     */
+    public function __construct(\MongoDB\Client $mongo, array $options = array(), float $initialTtl = 300.0)
+    {
+        if (!isset($options['database'])) {
+            throw new \InvalidArgumentException(
+                'You must provide the "database" option for MongoDBStore'
+            );
+        }
+
+        $this->mongo = $mongo;
+
+        $this->options = array_merge(array(
+            'collection' => 'lock',
+            'resource_field' => '_id',
+            'token_field' => 'token',
+            'acquired_field' => 'acquired_at',
+            'expiry_field' => 'expires_at',
+        ), $options);
+
+        $this->initialTtl = $initialTtl;
+    }
+
+    /**
+     * {@inheritdoc}
+     *
+     * db.lock.update(
+     *     {
+     *         _id: "test",
+     *         expires_at: {
+     *             $lte : new Date()
+     *         }
+     *     },
+     *     {
+     *         _id: "test",
+     *         token: {# unique token #},
+     *         acquired: new Date(),
+     *         expires_at: new Date({# now + ttl #})
+     *     },
+     *     {
+     *         upsert: 1
+     *     }
+     * );
+     */
+    public function save(Key $key)
+    {
+        $expiry = $this->createDateTime(microtime(true) + $this->initialTtl);
+
+        $filter = array(
+            $this->options['resource_field'] => (string) $key,
+            '$or' => array(
+                array(
+                    $this->options['token_field'] => $this->getToken($key),
+                ),
+                array(
+                    $this->options['expiry_field'] => array(
+                        '$lte' => $this->createDateTime(),
+                    ),
+                ),
+            ),
+        );
+
+        $update = array(
+            '$set' => array(
+                $this->options['resource_field'] => (string) $key,
+                $this->options['token_field'] => $this->getToken($key),
+                $this->options['acquired_field'] => $this->createDateTime(),
+                $this->options['expiry_field'] => $expiry,
+            ),
+        );
+
+        $options = array(
+            'upsert' => true,
+        );
+
+        $key->reduceLifetime($this->initialTtl);
+        try {
+            $this->getCollection()->updateOne($filter, $update, $options);
+        } catch (\MongoDB\Driver\Exception\BulkWriteException $e) {
```

Comment:
I've improved the exception handling, I also noticed `waitAndSave` was throwing an `InvalidArgumentException` instead of a `NotSupportedException`. The `RedisStore` also had this issue. I have fixed the `MondoDbStore` but we should also fix the `RedisStore`. I also adjusted the phpdoc in `StoreInterface` to include `LockExpiredException` in `save`, `waitAndSave` and `putOffExpiration` as per other store implementations.